Add Initial Mailbox panel and sensor

PhracturedBlue opened the pull request:
## Description:
This is a new panel and sensor that adds an interface to and Asterisk voice-mailbox

The panel allows viewing messages, reading a speech-to-text translation of the messages, listening to the messages, and deleting messages

The sensor shows how many messages are in the mailbox

**Note**
While the functionality is all in place, I am not fluent in working with asyncio.  I think I've fixed most of the data handlers now.
I believe the MP3 handler should now properly load asynchronously

I would really appreciate any code review, especially on the async aspects.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2908

**Pull request in [github.com/home-assistant/home-assistant-polymer) with frontend code change:** https://github.com/home-assistant/home-assistant-polymer/pull/319

## Example entry for `configuration.yaml` (if applicable):
```yaml
asterisk_mbox:
    password: This is my password
    host: localhost
    port: 12345
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

PhracturedBlue committed:
Initial implementation of Asterisk Mailbox

pvizeli reviewed (changes requested):
Before I start the review, some points to change:
- Don't store data into `hass.data`. You can use a own storage object an pass that into views. For sensors, they can get data from dispatcher without store into `hass.data`
- Remove unneeded code instead comment out
- Run `setup()` without async flavor. Sensor and view only access to ram cache or need to be async, they can be stay async.
- All client access inside views are not async and need run as Job
- Requerments need a pined version
- set depencencies to http component

PhracturedBlue committed:
Rework asterisk_mbox handler to avoid using the hass.data hash.  Fix requirements.

PhracturedBlue commented:
I think I have addressed most of your concerns, but I do not understand this one:
> set depencencies to http component

Can you clarify your request please?

PhracturedBlue committed:
Handle potential asterisk server disconnect.  bump asterisk_mbox requirement to 0.4.0

PhracturedBlue committed:
Use async method for mp3 fetch from server

balloob commented:
He means that you should add `DEPENDENCIES = ['http']` so that the HTTP component is loaded before your component in initialized. (otherwise you can't register your view)

PhracturedBlue committed:
Add http as dependency

PhracturedBlue committed:
Minor log fix. try to force Travis to rebuild

PhracturedBlue commented:
I have completed all of the development now, have completed the frontend impelementation and documentation, so I'm moving this out of WIP.  I will of course still appreciate any feedback.

PhracturedBlue commented:
I had a thought...Should I convert this into a generic 'mailbox' component?  The panel could easily be made a bit more generic and provide a list of any messages not just voicemail).

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,192 @@
+def setup(hass, config):
+    """Set up for the Asterisk Voicemail box."""
+    # pylint: disable=global-statement, import-error
+    global ASTERISK
```
You're not allowed to use globals. Sore the object in `hass.data`

PhracturedBlue replied:
In a different review, I was asked not to use hass.data[] to store data (or at least that was my interpretation of the request), so I did my best to avoid them here.

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,192 @@
+    discovery.load_platform(hass, "sensor", DOMAIN, {}, config)
+
+    if 'frontend' in hass.config.components:
+        register_built_in_panel(hass, 'mailbox', 'Mailbox',
```
Please use the new syntax: `hass.components.frontend.register_built_in_panel(…)` and remove the `hass` argument.

balloob commented:
Sorry for the delay responding to this PR. We've been busy wrapping up Hass.io.

I like the idea of making it into a generic mailbox component. That makes it easier to justify shipping a panel to all our users. I do think that mailbox might not be the best name, maybe voicemail ?

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,192 @@
+        """Retrieve Asterisk mp3."""
+        _LOGGER.info("Sending mp3 for %s", sha)
+
+        response = yield from self.handle(request, sha)
```
No real reason to create a handle method if there is only 1 function calling it?

PhracturedBlue replied:
I created the handle because I want to yield to it.  The callback to the asterisk server may take a significant time (hopefully not), and I didn't want to block other jobs while waiting for the reply.  I am not sure of a better way to address this.

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,192 @@
+    def handle(self, request, sha):
+        """Server MP3."""
+        client = self.data.client
+        hass = self.data.hass
```
```python
hass = request.app['hass']
client = hass.data[DOMAIN]
```

balloob replied:
That way you can remove the constructor and the instance variables.

PhracturedBlue commented:
I'll start work on the requested changes.  I called it 'mailbox' because there is very little inherent to voicemail in the panel (pretty much just the icon).  It could just as easily be email or twitter or list of camera photos, or whatever.  I thought that would make it more likely someone would devise an alternate backend.

PhracturedBlue committed:
Merge remote-tracking branch 'upstream/dev' into asterisk_mbox

PhracturedBlue committed:
Updates based on review

balloob commented:
Okay, we can leave it as mailbox.

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,171 @@
+            client = hass.data[DOMAIN].client
+            for sha in data:
+                _LOGGER.info("Deleting: %s", sha)
+                client.delete(sha)
```
Is this doing I/O ? If so, it should not be done inside an async method but be run inside the executor using `yield from hass.async_add_job(client.delete, sha)`

PhracturedBlue replied:
not directly no.  The I/O in this case happens in a separate thread in the asterisk_mbox module.  all this function does is push a command onto the queue and returns.  This is different than the mp3 command which has to wait for an I/O response.

I will work on refactoring the code to create a 'mailbox' component  of which the asterisk_mbox will be a member.

balloob commented on homeassistant/components/asterisk_mbox.py:
```diff
@@ -0,0 +1,171 @@
+
+        with suppress(asyncio.CancelledError, asyncio.TimeoutError):
+            with async_timeout.timeout(10, loop=request.app['hass'].loop):
+                stream = yield from hass.async_add_job(fetch)
```
This won't work, as you're going to return `self.json_message` if something goes wrong.

Since exceptions propagate, you can just do 

```python
from functools import partial

try:
    stream = yield from hass.async_add_job(partial(client.mp3, sha, sync=True))
except ServerError as err:
```

PhracturedBlue committed:
Fix error handling as per review

PhracturedBlue committed:
Fix error handling as per review

PhracturedBlue committed:
Refactor voicemail into mailbox component

PhracturedBlue commented:
I have refactored the code now.  I would appreciate it if you could re-review to make sure the new structure is acceptable.  In the mean time, I am going to try to develop a demo component and then some tests for it.

PhracturedBlue committed:
Hide mailbox component from front page

PhracturedBlue committed:
Add demo for mailbox

PhracturedBlue committed:
Add tests for mailbox

PhracturedBlue committed:
Remove asterisk_mbox sensor and replace with a generic mailbox sensor

PhracturedBlue committed:
Fix linting errors

balloob commented on tests/components/mailbox/test_init.py:
```diff
@@ -0,0 +1,158 @@
+    get_test_home_assistant, get_test_instance_port, assert_setup_component)
+
+
+class TestSetupMailbox(object):
```
You can rewrite this whole test class as a test function on the root level. Use `hass` as a parameter to the test function to automatically get a new test instance:

```python
def test_setup_component(has):
    """Initialize demo platform on mailbox component."""
    config = {
        mailbox.DOMAIN: {
            'platform': 'demo'
        }
    }

    with assert_setup_component(1, mailbox.DOMAIN):
        setup_component(hass, mailbox.DOMAIN, config) 
```

balloob commented on tests/components/mailbox/test_init.py:
```diff
@@ -0,0 +1,158 @@
+            }
+        }
+
+        setup_component(self.hass, mailbox.DOMAIN, config)
```
You always want to `assert` the result of `setup_component`. That way it will fail the test early if setup fails.

PhracturedBlue commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,175 @@
+    @property
+    def hidden(self) -> bool:
+        """Return whether the component should be hidden."""
+        return True
```
To clarify, are you saying that I should have the mailbox sensor be initialized separately with a parameter that indicates which platform it is associated with?  then the mailbox is exclusively a platform, and the mailbox sensor is the entity.  Can a platform hold a state (I thought that was only an entity thing).  I don't understand how to have the sensor be notified of new messages without duplicating all of the platform code in the sensor if I do it that way.  Today the sensor follow the mailbox, so all of the business logic resides in the mailbox.  I can use message passing to broadcast message changes I suppose? I also am not sure how to identify the available platforms from the frontend without a state.  Should I create an api that the frontend queries to ask about available platforms?

balloob replied:
You will have to remove the mailbox sensor completely. Components are not allowed to have sensors in other platforms that forward other platform data. 

If you still want to have an entity to write the current unread messages to the state machine, then that's a responsibility of the mailbox component to create entities like `mailbox.platform_name` with as state the number of unread messages.

However, you cannot combine the entity and fetching of messages in 1 object. You can have the entity poll the service object for new messages and then update the state.

PhracturedBlue replied:
Ok.  what about querying the available platforms from the frontend? Do I just create an API call for that?

Also, what about identifying when the messages have changed on the frontend?  Today I can query the entity state object for that.  If the frontend can't rely on the 'last_updated' field of the entiity, do I just need to use polling here?  It seems less elegant than being informed when a change has happened automatically.

balloob replied:
You can just fire events when that happens. Have a look at how the shopping list works.

PhracturedBlue replied:
Okay, I've switched over to using events now.

PhracturedBlue committed:
Remove mailbox sensor.  Remove demo.mp3.  Split entity from platform object.

PhracturedBlue committed:
Update mailbox test

PhracturedBlue committed:
Update mailbox test

PhracturedBlue commented:
I reworked the mailbox to separate the entity from the platform object and removed the sensor.  I am not precisely sure if this is more aligned with what you were looking for though.

PhracturedBlue committed:
Use events to indicate state change rather than entity last-updated

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+        for mailbox in self.mailboxes:
+            if mailbox.name == platform:
+                return mailbox
+        return None
```
Raise a 404 directly so you don't have to check for `None`

```
from aiohttp.web_exceptions import HTTPNotFound

raise HTTPNotFound
```

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+class MailboxDeleteView(MailboxView):
+    """View to delete selected messages."""
+
+    url = "/api/mailbox/delete/{platform}"
```
If you want this to be restful:

url = "/api/mailbox/{platform}/{message_id}"

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+    name = "api:mailbox:delete"
+
+    @asyncio.coroutine
+    def post(self, request, platform):
```
This should be `delete(…)`

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+        if mailbox is None:
+            return web.Response(status=401)
+        try:
+            data = yield from request.json()
```
Any JSON received from the client should be validated before being passed on. But if you go restful (as I commented), you won't need this.

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+        self.hass = hass
+        self.name = name
+
+    def update(self):
```
All these methods are getting called from async contexts. Please prefix all names with `async`

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+        """Send event notification of updated mailbox."""
+        self.hass.bus.async_fire(EVENT)
+
+    def get_media_type(self):
```
Fetching data from any platform should be tagged as `@asyncio.coroutine`.

balloob commented on homeassistant/components/mailbox/__init__.py:
```diff
@@ -0,0 +1,256 @@
+            with async_timeout.timeout(10, loop=hass.loop):
+                try:
+                    stream = yield from hass.async_add_job(
+                        partial(mailbox.get_media, msgid))
```
Don't use partial, it messes with the annotations and ends up running everything in the executor. async_add_job already supports passing arguments as extra arguments:

```python
yield from hass.async_add_job(mailbox.get_media, msgid)
```

balloob commented on tests/components/mailbox/test_init.py:
```diff
@@ -0,0 +1,148 @@
+
+        setup_component(
+            self.hass, http.DOMAIN,
+            {http.DOMAIN: {http.CONF_SERVER_PORT: get_test_instance_port()}})
```
Please do not use `get_test_instance_port`. We deprecated the use of them.

Have a look at the API tests for example how to write tests that consume HTTP views https://github.com/home-assistant/home-assistant/blob/dev/tests/components/test_api.py

That file is also a good example showing how you should not use a test class anymore but instead use functions at the root level to represent tests.

balloob commented:
Yes, the current architecture is great.

PhracturedBlue committed:
Make mailbox platform calls async.  Fix other review concerns

PhracturedBlue committed:
Rewrite mailbox tests to live at root level and be async.  Fixmailbox dependency on http

PhracturedBlue committed:
Only store number of messages not content in mailbox entity

PhracturedBlue commented:
One thing I wasn't sure about was whether I was allowed to use a '@property' tag on the platform object.  the 'media_type' value is static (per platform) and it seemed excessive to use an coroutine to retrieve it.  I had only seen '@prperty' on entities though, so I wasn't sure it is ok.

balloob commented:
Yeah, properties are fine . If it's static it doesn't even have to be a property per se. Just this could work too:

```python
class Bla:
  media_type = 'something'
```

balloob commented:
🎉  🐬  Awesome, great work!
